bluesky_env/devices/top_align.py

```diff
@@ -0,0 +1,343 @@
+    pass
+
+@register_plugin
+class CVPlugin(PluginBase):
+    _default_suffix = "CV1:"
+    _suffix_re = "CV1\d:"
+    _default_read_attrs = ["outputs"]
+    func_sets = DDC_EpicsSignal(
+        *[(f"func_set{k}", f"CompVisionFunction{k}") for k in range(1, 4)]
+    )
+    inputs = DDC_EpicsSignal(
+        *[(f"input{k}", f"Input{k}") for k in range(1, 11)]
+    )
+    outputs = DDC_EpicsSignalRO(
+        *[(f"output{k}", f"Output{k}_RBV") for k in range(1, 11)]
+    )
+    cam_depth = ADCpt(EpicsSignal, "CompVisionCamDepth", kind="config")
+
+
+class StandardProsilica(SingleTrigger, ProsilicaDetector):
+    image = Cpt(ImagePlugin, "image1:")
+    roi1 = Cpt(ROIPlugin, "ROI1:")
+    roi2 = Cpt(ROIPlugin, "ROI2:")
+    roi3 = Cpt(ROIPlugin, "ROI3:")
+    roi4 = Cpt(ROIPlugin, "ROI4:")
+    trans1 = Cpt(TransformPlugin, "Trans1:")
+    proc1 = Cpt(ProcessPlugin, "Proc1:")
+    stats1 = Cpt(StatsPlugin, "Stats1:")
+    stats2 = Cpt(StatsPlugin, "Stats2:")
+    stats3 = Cpt(StatsPlugin, "Stats3:")
+    stats4 = Cpt(StatsPlugin, "Stats4:")
+    stats5 = Cpt(StatsPlugin, "Stats5:")
+
+class EpicsMotorSPMG(EpicsMotor):
+    SPMG = Cpt(EpicsSignal, ".SPMG")
+
+
+class TopAlignCam(StandardProsilica):
+    _default_read_attrs = ["cv1", "tiff"]
+    cv1 = Cpt(CVPlugin, "CV1:")
+    tiff = Cpt(
+        TIFFPluginWithFileStore,
+        "TIFF1:",
+        write_path_template="/nsls2/data/amx/legacy/topcam",
+    )
+    cam_mode = Cpt(Signal, value=None, kind="config")
+    pix_per_um = Cpt(Signal, value=0.164, doc="pixels per um")
+    out10_buffer = Cpt(
+        EpicsSignalRO,
+        "Out10:compress",
+        doc="circular buffer for ADCompVision output10 (thresholding loop profile)",
+    )
+    out10_reset = Cpt(EpicsSignal, "Out10:compress.RES")
+    out9_buffer = Cpt(EpicsSignalRO, "Out9:compress")
+    out9_reset = Cpt(EpicsSignal, "Out9:compress.RES")
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(
+            *args,
+            **kwargs,
+        )
+        #self.tiff.read_attrs = []
+        self.cv1.read_attrs = ["outputs"]
+        self.cv1.outputs.read_attrs = [
+            "output8",
+            "output9",
+            "output10",
+        ]
+        self.cam_mode.subscribe(self._update_stage_sigs, event_type="value")
+
+    def _update_stage_sigs(self, *args, **kwargs):
+        self.tiff.stage_sigs.update([("enable", 0)])
+        self.stage_sigs.clear()
+        self.stage_sigs.update(
+            [
+                ("cv1.func_sets.func_set1", "Canny Edge Detection"),
+                ("cv1.inputs.input1", 8),
+                ("cv1.inputs.input2", 3),
+                ("cv1.inputs.input3", 1),
+                ("cv1.inputs.input4", 3),
+                ("cv1.inputs.input5", 13),
+                ("cv1.inputs.input6", 1),
+                ("trans1.nd_array_port", "PROC1"),
+                ("tiff.nd_array_port", "CV1"),
+            ]
+        )
+        if self.cam_mode.get() == "coarse_align":
```

Comment:
some docs or other info about the `cam_mode`s, as well as having constants for the strings would help clean up the code and reduce issues with setting the values.

Reply:
Added enum for cam modes and doc strings in 1608ff2f2a52dea4959c622a2f9cd604414f0f35